### PR TITLE
CORDA-3725 Fix SQL deadlocks coming from soft locking states

### PR DESCRIPTION
**Pull request by kyriathar**
Fix SQL deadlocks coming from soft locking states;

Adding to the query -explicitly- the flow's locked states resolved the SQL Deadlocks in SQL server. The SQL Deadlocks would come up  from `softLockRelease` when only the `lockId` was passed in as argument. In that case the query optimizer would use `lock_id_idx(lock_id, state_status)` to search and update entries in `VAULT_STATES` table. However, all rest of the queries would follow the opposite direction meaning they would use PK's `index(output_index, transaction_id)` but they would also update the `lock_id` column and therefore the `lock_id_idx` as well, because `lock_id` is a part of it. That was causing a circular locking among the different transactions (SQL processes) within the database.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt [hunk 1]
```diff
@@ -90,15 +91,18 @@ class ActionExecutorImpl(
...
         val serializedFlowState = when(flowState) {
             FlowState.Completed -> null
+            // upon implementing CORDA-3816: If we have errored or hospitalized then we don't need to serialize the flowState as it will not get saved in the DB
             else -> flowState.checkpointSerialize(checkpointSerializationContext)
         }
+        // upon implementing CORDA-3816: If we have errored or hospitalized then we don't need to serialize the serializedCheckpointState as it will not get saved in the DB
+        val serializedCheckpointState: SerializedBytes<CheckpointState> = checkpoint.checkpointState.checkpointSerialize(checkpointSerializationContext)
         if (action.isCheckpointUpdate) {
-            checkpointStorage.updateCheckpoint(action.id, checkpoint, serializedFlowState)
+            checkpointStorage.updateCheckpoint(action.id, checkpoint, serializedFlowState, serializedCheckpointState)
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt]: Is this related to soft locking?  Just wondering how these changes are part of the PR and not already in the feature branch ?

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt]: Oh, this is because we made some changes to serialization to make it use checkpoint serializer ?

**Review by rick-r3 — COMMENTED** [review]
I think this looks good.  @lankydan should have a look, and of course, we should squash all this down when merging it into the feature branch.

File: node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt (hunk 1 above)
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt]: ye these are part of the other PR and its because he has branched off from it. We talked about sorting this PR out when the other is in.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 2]
```diff
@@ -42,9 +42,29 @@ import javax.persistence.criteria.CriteriaUpdate
 import javax.persistence.criteria.Predicate
 import javax.persistence.criteria.Root
 
-private fun CriteriaBuilder.executeUpdate(session: Session, configure: Root<*>.(CriteriaUpdate<*>) -> Any?) = createCriteriaUpdate(VaultSchemaV1.VaultStates::class.java).let { update ->
-    update.from(VaultSchemaV1.VaultStates::class.java).run { configure(update) }
-    session.createQuery(update).executeUpdate()
+private fun CriteriaBuilder.executeUpdate(
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: As we are changing this piece of code can we move it from here and into the class itself. Probably just put it at the bottom of the class.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 3]
```diff
@@ -42,9 +42,29 @@ import javax.persistence.criteria.CriteriaUpdate
...
+    session: Session,
+    stateRefs: NonEmptySet<StateRef>?,
+    configure: Root<*>.(CriteriaUpdate<*>, List<PersistentStateRef>?) -> Any?
+): Int {
+    fun doUpdate(persistentStateRefs: List<PersistentStateRef>?): Int {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Should `doUpdate` be its own function?

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 4]
```diff
@@ -42,9 +42,29 @@ import javax.persistence.criteria.CriteriaUpdate
...
+        createCriteriaUpdate(VaultSchemaV1.VaultStates::class.java).let { update ->
+            update.from(VaultSchemaV1.VaultStates::class.java).run { configure(update, persistentStateRefs) }
+            return session.createQuery(update).executeUpdate()
+        }
+    }
+    return stateRefs?.let {
+        // divide [List<PersistentStateRef>] into [List<List<PersistentStateRef>>] of [MAX_SQL_IN_CLAUSE_SET] size each for sql server performance
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Change comment to:

```
// Increase SQL server performance by, processing updates in chunks allowing the database's optimizer to use the table's index.
```

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 5]
```diff
@@ -508,37 +535,60 @@ class NodeVaultService(
...
             val lockIdPredicate = criteriaBuilder.equal(get<String>(VaultSchemaV1.VaultStates::lockId.name), lockId.toString())
             update.set<String>(get<String>(VaultSchemaV1.VaultStates::lockId.name), criteriaBuilder.nullLiteral(String::class.java))
             update.set(get<Instant>(VaultSchemaV1.VaultStates::lockUpdateTime.name), softLockTimestamp)
-            configure(update, arrayOf(stateStatusPredication, lockIdPredicate))
+            configure(update, arrayOf(stateStatusPredication, lockIdPredicate), persistentStateRefs)
         }
-        if (stateRefs == null) {
-            val update = execute { update, commonPredicates ->
+
+        // CORDA-3725: Adding to the query explicitly the flow's locked states resolved the SQL Deadlocks in sql server.
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Are these comments needed? Is this PR not resolving the issue stated in the comments?

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 6]
```diff
@@ -508,37 +535,60 @@ class NodeVaultService(
...
+        // The SQL Deadlocks were caused from softLockRelease when only the lockId was specified. In that case the query optimizer
+        // would use lock_id_idx(lock_id, state_status) to search and update entries in vault_states table. However, all rest of the queries would follow the
+        // opposite direction meaning they would use PK's index(output_index, transaction_id) but they could also update the lock_id and therefore the lock_id_idx as well.
+        // That was causing a circular locking among the different transactions within the database.
+        val flowStateMachineImpl = FlowStateMachineImpl.currentStateMachine()
+        val stateRefsToBeReleased =
+            stateRefs ?: flowStateMachineImpl?.let {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: If we are going to do a `let` we might as well remove the variable and directly call `FlowStateMachineImpl.currentStateMachine()`.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 7]
```diff
@@ -508,37 +535,60 @@ class NodeVaultService(
...
+                    flowStateMachineImpl?.let {
+                        if (lockId == it.id.uuid) {
+                            it.softLockedStates.removeAll(stateRefsToBeReleased)
+                        }
+                    }
+                    log.trace { "Releasing $updatedRows soft locked states for $lockId and stateRefs $stateRefsToBeReleased" }
                 }
             } catch (e: Exception) {
-                log.error("""soft lock update error attempting to release states for $lockId and $stateRefs")
+                log.error("""soft lock update error attempting to release states for $lockId and $stateRefsToBeReleased")
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Not your fault, but this log line looks screwed up. Should it not be:

```kotlin
log.error("Soft lock update error attempting to release states for $lockId and $stateRefsToBeReleased", e)
```

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 8]
```diff
@@ -508,37 +535,60 @@ class NodeVaultService(
         }
     }
 
+    @Suppress("NestedBlockDepth", "ComplexMethod")
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:531]: Add comments to this and the add method that gives some sort of idea to what the implementation is doing.

A.k.a, that if the locked states are locked with a flows id that they are held in the flow's fiber (in-memory) otherwise they are handled by the database.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 9]
```diff
@@ -831,6 +848,68 @@ class FlowFrameworkTests {
         assertEquals(null, persistedException)
     }
 
+    @Test(timeout=300_000)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Maybe we should consider moving the soft lock tests to their own test class.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 6 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: I only added the local variable because we also using it further down when, `if (updatedRows > 0) { ... softLockedStates.removeAll...`  although yeah its impossible that we are gonna lose the flowStateMachineImpl object in between since we running code on it.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 10]
```diff
@@ -1204,4 +1283,111 @@ internal class SuspendingFlow : FlowLogic<Unit>() {
...
+        }
+
+        @Suspendable
+        override fun call(): Boolean {
+            val unlockedStatesSize = unlockedStates.size
+            val emptySet = emptySet<StateRef>()
+            val lockSetFlowId = unlockedStates.subList(0, unlockedStatesSize / 2).map { it.ref }.toNonEmptySet()
+            val lockSetRandomId = unlockedStates.subList(unlockedStatesSize / 2, unlockedStatesSize).map { it.ref }.toNonEmptySet()
+
+            // lock and release with our flow Id
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Can we split the testing of locking by flow id into a separate test? We should have one that tests with a random id + the flow id like the test is already doing, but we should also test them in isolation.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 6 above)
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Oh i didn't see it down there, it can be left if you want then.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 3 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: I think, for now, since we only using it from within the method enclosing it, then it is maybe ok to stay in there, not entirely sure though.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 4 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Slightly modifying suggestion  to:

```
// Increase SQL server performance by, processing updates in chunks allowing the database's optimizer to make use of the index.
```

to be a bit more generic i think, if that's ok.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 5 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Indeed PR's description is same with this comment. Removing the comment.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 6 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Removing `val flowStateMachineImpl` as it adds noise.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 7 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Done -changed to proposed log line-.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 2 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Done -moved it at the bottom of the file-.

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 8 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:531]: Done -added comments for `softLockReserve` and `softLockRelease`-.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 9 above)
Comment (kyriathar) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done -tests and testing flows/ functions related were moved to delicated file-.

**Review by rick-r3 — DISMISSED** [review]
I'm happy if you're happy @lankydan

**Comment by kyriathar** [comment]
There is one final refactoring to tests left to be pushed @rick-r3 .

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 10 above)
Comment (kyriathar) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done -refactored testing flow to be configurable which allowed to break existing tests into multiple sub tests and added more tests-.

**Review by lankydan — APPROVED** [review]
